web_src/js/features/websocket.sharedworker.js

```diff
@@ -0,0 +1,172 @@
+const sourcesByUrl = {};
+const sourcesByPort = {};
+
+class Source {
+  constructor(url) {
+    this.url = url.replace(/^http/, 'ws');
+    this.webSocket = new WebSocket(this.url);
+    this.listening = {};
+    this.clients = [];
+    this.listen('open');
+    this.listen('close');
+    this.listen('logout');
+    this.listen('notification-count');
+    this.listen('stopwatches');
+    this.listen('error');
+    this.webSocket.addEventListener('error', (error) => {
+      this.lastError = error;
+    });
+    this.webSocket.addEventListener('message', (event) => {
+      const message = JSON.parse(event.data);
+      if (!message) {
+        return;
+      }
+      if (this.listening[message.Name]) {
+        this.notifyClients({
+          type: message.Name,
+          data: message.Data
+        });
+      }
+    });
+    this.webSocket.addEventListener('close', (event) => {
+      if (!this.webSocket) {
+        return;
+      }
+      const oldWebSocket = this.webSocket;
+      this.webSocket = null;
+      this.notifyClients({
+        type: 'close',
+        data: event
+      });
+      oldWebSocket.close();
+    });
+  }
+
+  register(port) {
+    if (this.clients.includes(port)) return;
+
+    this.clients.push(port);
+
+    port.postMessage({
+      type: 'status',
+      message: `registered to ${this.url}`,
+    });
+
+    if (!this.webSocket) {
+      if (this.lastError) {
+        port.postMessage({
+          type: 'error',
+          message: `websocket disconnected: ${this.lastError}`
+        });
+      } else {
+        port.postMessage({
+          type: 'error',
+          message: 'websocket disconnected'
+        });
+      }
+    }
+  }
+
+  deregister(port) {
+    const portIdx = this.clients.indexOf(port);
+    if (portIdx < 0) {
+      return this.clients.length;
+    }
+    this.clients.splice(portIdx, 1);
+    return this.clients.length;
+  }
+
+  close() {
+    if (!this.webSocket) return;
+    const oldWebSocket = this.webSocket;
+    this.webSocket = null;
+    oldWebSocket.close();
```

Comment:
For resilience on flaky connections, we want to reconnect on abnormal closure (should be `e.code >= 1001 && event.code < 3999`), after a small timeout of like 1s, ideally giving up after a certain period of failure, like 30s.